quint/src/runtime/impl/compilerImpl.ts

```diff
@@ -387,514 +392,519 @@ export class CompilerVisitor implements IRVisitor {
-        break
-
-      case 'iadd':
-        this.applyFun(app.id, 2, (p, q) => just(rv.mkInt(p.toInt() + q.toInt())))
-        break
-
-      case 'isub':
-        this.applyFun(app.id, 2, (p, q) => just(rv.mkInt(p.toInt() - q.toInt())))
-        break
-
-      case 'imul':
-        this.applyFun(app.id, 2, (p, q) => just(rv.mkInt(p.toInt() * q.toInt())))
-        break
-
-      case 'idiv':
-        this.applyFun(app.id, 2, (p, q) => {
-          if (q.toInt() !== 0n) {
-            return just(rv.mkInt(p.toInt() / q.toInt()))
-          } else {
-            this.errorTracker.addRuntimeError(app.id, 'Division by zero')
-            return none()
-          }
-        })
-        break
-
-      case 'imod':
-        this.applyFun(app.id, 2, (p, q) => just(rv.mkInt(p.toInt() % q.toInt())))
-        break
-
-      case 'ipow':
-        this.applyFun(app.id, 2, (p, q) => {
-          if (q.toInt() == 0n && p.toInt() == 0n) {
-            this.errorTracker.addRuntimeError(app.id, '0^0 is undefined')
-          } else if (q.toInt() < 0n) {
-            this.errorTracker.addRuntimeError(app.id, 'i^j is undefined for j < 0')
-          } else {
-            return just(rv.mkInt(p.toInt() ** q.toInt()))
-          }
-          return none()
-        })
-        break
-
-      case 'igt':
-        this.applyFun(app.id, 2, (p, q) => just(rv.mkBool(p.toInt() > q.toInt())))
-        break
-
-      case 'ilt':
-        this.applyFun(app.id, 2, (p, q) => just(rv.mkBool(p.toInt() < q.toInt())))
-        break
-
-      case 'igte':
-        this.applyFun(app.id, 2, (p, q) => just(rv.mkBool(p.toInt() >= q.toInt())))
-        break
-
-      case 'ilte':
-        this.applyFun(app.id, 2, (p, q) => just(rv.mkBool(p.toInt() <= q.toInt())))
-        break
-
-      case 'Tup':
-        // Construct a tuple from an array of values
-        this.applyFun(app.id, app.args.length, (...values: RuntimeValue[]) => just(rv.mkTuple(values)))
-        break
-
-      case 'item':
-        // Access a tuple: tuples are 1-indexed, that is, _1, _2, etc.
-        this.applyFun(app.id, 2, (tuple, idx) => this.getListElem(app.id, tuple.toList(), Number(idx.toInt()) - 1))
-        break
-
-      case 'tuples':
-        // Construct a cross product
-        this.applyFun(app.id, app.args.length, (...sets: RuntimeValue[]) => just(rv.mkCrossProd(sets)))
-        break
-
-      case 'List':
-        // Construct a list from an array of values
-        this.applyFun(app.id, app.args.length, (...values: RuntimeValue[]) => just(rv.mkList(values)))
-        break
-
-      case 'range':
-        this.applyFun(app.id, 2, (start, end) => {
-          const [s, e] = [Number(start.toInt()), Number(end.toInt())]
-          if (s <= e) {
-            const arr: RuntimeValue[] = []
-            for (let i = s; i < e; i++) {
-              arr.push(rv.mkInt(BigInt(i)))
+    if (!ir.isQuintBuiltin(app)) {
+      this.applyUserDefined(app)
+    } else {
```

Comment:
The only change here was adding a conditional using a type guard to check whether we have a builtin before dispatching the builtins.